DBZ-5283 Introduce a new SMT ExtractChangedRecordState to capture the names of changed/unchanged field within headers (#3980)
debezium/debezium: documentation/modules/ROOT/pages/transformations/event-changes.adoc

[source,json,indent=0]
----
{
	"op": "u",
	"source": {
		...
	},
	"ts_ms" : "...",
	"before" : {
		"field1" : "oldvalue1",
		"field2" : "oldvalue2"
	},
	"after" : {
		"field1" : "newvalue1",
		"field2" : "newvalue2"
	}
}
----

ifdef::community[]
More details about change event structure are provided in
xref:{link-connectors}[the documentation for each connector].
endif::community[]

The complex format of the message in the preceding example provides detailed information about changes that occur in the source database.
However, the format might not be suitable for some downstream consumers.
Sink connectors, or other parts of the Kafka ecosystem might expect the message to explicitly identify the fields that a database operation changes or leaves unchanged.
The `ExtractChangedRecordState` SMT adds headers to the change event message to identify the fields that are modified by a database operation, and the fields that remain unchanged.

[[event-changes-behavior]]
== Behavior

The event changes SMT extracts the `before` and `after` fields from a {prodname} `UPDATE` change event in a Kafka record.

> The SMT examines the fields in the `before` and `after` event state structures, and then writes the names of fields that changed or are unchanged based on the values of  settings that you apply to the SMT configuration.

Reviewer: ```suggestion
The transformation examines the `before` and `after` event state structures to identify the fields that are altered by an operation, and those that remain unchanged.
Depending on the connector configuration, the transformation then produces a modified event message that adds message headers to list the changed fields, the unchanged fields, or both.
```

Reviewer: @Naros A few small wording changes to further unpack the meaning. If this looks ok to you, I'm good with releasing this to the wild.

If the event represents an `INSERT` or `DELETE`, this single message transformation has no effect.

You can configure the event changes SMT for a {prodname} connector, or for a sink connector that consumes messages emitted by a {prodname} connector.
Configure the event changes SMT for a sink connector if you want Apache Kafka to retain the entire original {prodname} change events.
The decision to apply the SMT to a source or sink connector depends on your particular use case.

Depending on your use case, you can configure the transformation to modify the original message by performing one or both of the following tasks:

* Identify the fields that are changed by an `UPDATE` event by listing them in the user-configured xref:extract-changes-header-changed-name[`header.changed.name`] header.
* Identify the fields that are not changed by an `UPDATE` event by listing them in the user-configured xref:extract-changes-header-unchanged-name[`header.unchanged.name`] header.

== Configuration

Configure the {prodname} event changes SMT in a Kafka Connect source or sink connector by adding the SMT configuration details to your connector's configuration.

> To obtain the default behavior which doesn't add any headers, in a `.properties` file, you would specify something like the following:

Reviewer 2: I don't think we should talk about `.properties` file. It is connector configuration.

Reviewer: This is copied in from something that I introduced in some of the other transformation docs a couple of years ago. It appears in most of the other docs. We can remove it here and on L103 @Naros, and I'll open a Jira to correct instances that appear elsewhere.

Reviewer 3: I believe these are just copied from the `ExtractNewRecordState` SMT documentation and amended to be relevant to this SMT, but the formatting is relatively the same.  I see Bob has provided some edits below, I'll adjust accordingly.

> To obtain the default behavior which doesn't add any headers, in a `.properties` file, you would specify something like the following:

Reviewer: Removes reference on L93 to `.properties` file per earlier comment.

[source]
----
transforms=changes,...
transforms.changes.type=io.debezium.transforms.ExtractChangedRecordState
----

As with any Kafka Connect connector configuration, you can set `transforms=` to multiple, comma-separated, SMT aliases in the order in which you want Kafka Connect to apply the SMTs.

> The following `.properties` example sets several event changes SMT options:

Reviewer 2: See above

> The following `.properties` example sets several event changes SMT options:

Reviewer: ```suggestion
The connector configuration in the following example sets several options for the event changes SMT:
```

Reviewer: Removes reference on L103 to `.properties` file per earlier comment.

== Options for applying the event changes transformation selectively

In addition to the change event messages that a {prodname} connector emits when a database change occurs, the connector also emits other types of messages, including heartbeat messages, and metadata messages about schema changes and transactions.
Because the structure of these other messages differs from the structure of the change event messages that the SMT is designed to process, it's best to configure the connector to selectively apply the SMT, so that it processes only the intended data change messages.

For more information about how to apply the SMT selectively, see xref:{link-smt-predicates}#applying-transformations-selectively[Configure an SMT predicate for the transformation].

[id="configuration-options"]

== Configuration options

The following table describes the options that you can specify to configure the event changes SMT.

.Descriptions of event changes SMT configuration options
[cols="30%a,25%a,45%a",subs="+attributes",options="header"]
|===
|Option
|Default
|Description

|[[extract-changes-header-changed-name]]<<extract-changes-header-changed-name, `+header.changed.name+`>>
|

> |Specifies the name of the header that will include a comma-separated list of fields that were changed.

Reviewer: ```suggestion
|The Kafka message header name to use for storing a comma-separated list of the fields that are changed by a database operation.
```

|[[extract-changes-header-unchanged-name]]<<extract-changes-header-unchanged-name, `+header.unchanged.name+`>>
|

> |Specifies the name of the header that will include a comma-separated list of fields that were not changed.

Reviewer: ```suggestion
|The Kafka message header name to use for storing a comma-separated list of the fields that remain unchanged after a database operation.
```

The decision to apply the SMT to a source or sink connector depends on your particular use case.

You can configure the transformation to do any of the following:

* Identify the fields that are changed by an `UPDATE` event by listing them in the user-configured xref:extract-changes-header-changed-name[`header.changed.name`] header.
* Identify the fields that are not changed by an `UPDATE` event by listing them in the user-configured xref:extract-changes-header-unchanged-name[`header.unchanged.name`] header.

The single message transformation can be configured to include both changed and unchanged fields or either, depending on your use case.

== Configuration

Configure the {prodname} event changes SMT in a Kafka Connect source or sink connector by adding the SMT configuration details to your connector's configuration.
To obtain the default behavior which doesn't add any headers, in a `.properties` file, you would specify something like the following:

[source]
----
transforms=changes,...
transforms.changes.type=io.debezium.transforms.ExtractChangedRecordState
----

> As with any Kafka Connect connector configuration, you can set `transforms=` to multiple, comma-seperated, SMT aliases in the order in which you want Kafka Connect to apply the SMTs.

Reviewer: ```suggestion
As with any Kafka Connect connector configuration, you can set `transforms=` to multiple, comma-separated, SMT aliases in the order in which you want Kafka Connect to apply the SMTs.
```

:page-aliases: configuration/event-changes.adoc
[id="event-record-changes"]
= Event Record Changes

:toc:
:toc-placement: macro
:linkattrs:
:icons: font
:source-highligher: highlight.js

toc::[]

ifdef::community[]
[NOTE]
====
This single message transformation (SMT) is supported for only the SQL database connectors.
====
endif::community[]

A {prodname} data change event has a complex structure that provides a wealth of information.
However, parts of the chagne data capture pipeline may require knowledge of what fields were changed or unchanged when the event occurred in the source system.
To provide this information, {prodname} provides a event record changes single message transformation (SMT).

The event changes transformation is a
link:https://kafka.apache.org/documentation/#connect_transforms[Kafka Connect SMT].

== Change event structure

{prodname} generates data change events that have a complex structure.
Each event consists of three parts:

* Metadata, which includes but is not limited to:

> ** The operation that made the change
> ** Source information such as the names of the database and table where the change was made
> ** Time stamp for when the change was made

Reviewer: ```suggestion
** The type of operation that changed the data.
** Source information, such as the names of the database and the table in which the change occurred.
** Timestamp that identifies when the change was made.
```

Reviewer: I know that this same list appears elsewhere (e.g., the event-flattening SMT docs). I'm in the process of making similar changes there as well.

** Optional transaction information

* Row data before the change
* Row data after the change

> For example, part of the structure of an `UPDATE` change event looks like this:

Reviewer: ```suggestion
The following example shows part of the structure of a typical {prodname} `UPDATE` change event:
```

[source,json,indent=0]
----
{
	"op": "u",
	"source": {
		...
	},
	"ts_ms" : "...",
	"before" : {
		"field1" : "oldvalue1",
		"field2" : "oldvalue2"
	},
	"after" : {
		"field1" : "newvalue1",
		"field2" : "newvalue2"
	}
}
----

ifdef::community[]
More details about change event structure are provided in
xref:{link-connectors}[the documentation for each connector].
endif::community[]

This complex format provides the most information about changes happening in the system.
However, other connectors or other parts of the Kafka ecosystem may expect knowledge of what fields changed or are unchanged, and this event changes SMT allows exposing that information via event headers.

[[event-changes-behavior]]
== Behavior

The event changes SMT extracts the `before` and `after` fields from a {prodname} `UPDATE` change event in a Kafka record.
The SMT will examine each of the fields within these two  event state structures and writes the names of fields that changed or are unchanged based on the configuration supplied to the SMT.
If the event represents an `INSERT` or `DELETE`, this single message transformation has no effect.

> You can configure the event changes SMT for a {prodname} connector or for a sink connector that consumes messages emitted by a {prodname} connector.

Reviewer: ```suggestion
You can configure the event changes SMT for a {prodname} connector, or for a sink connector that consumes messages emitted by a {prodname} connector.
```

> The advantage of configuring the event changes for a sink connector is that the records stored in Apache Kafka contain whole {prodname} change events.

Reviewer: ```suggestion
Configure the event changes SMT for a sink connector if you want Apache Kafka to retain the entire original {prodname} change events .
```

The decision to apply the SMT to a source or sink connector depends on your particular use case.

You can configure the transformation to do any of the following:

> * Record the fields that changed in the `UPDATE` in a user-configured header.
> * Record the fields that did not change in the `UPDATE` in a user-configured header.

Reviewer: ```suggestion
* Identify the fields that are changed by an `UPDATE` event by listing them in the user-configured xref:extract-changes-header-changed-name[`header.changed.name`] header.
* Identify the fields that are not changed by an `UPDATE` event by listing them in the user-configured xref:extract-changes-header-unchanged-name[`header.unchanged.name`] header.
```

xref:{link-connectors}[the documentation for each connector].
endif::community[]

The complex format of the message in the preceding example provides detailed information about changes that occur in the source database.
However, the format might not be suitable for some downstream consumers.
Sink connectors, or other parts of the Kafka ecosystem might expect the message to explicitly identify the fields that a database operation changes or leaves unchanged.

> The `ExtractChangedRecordState` SMT adds headers to the change event message to identify the fields that are modified by a database operation or left unchanged.

Reviewer: ```suggestion
The `ExtractChangedRecordState` SMT adds headers to the change event message to identify the fields that are modified by a database operation, and the fields that remain unchanged.
```

Configure the event changes SMT for a sink connector if you want Apache Kafka to retain the entire original {prodname} change events.
The decision to apply the SMT to a source or sink connector depends on your particular use case.

> Depending on your use case, you can configure the transformation to modify the original message by performing one of both of the following tasks:

Reviewer: ```suggestion
Depending on your use case, you can configure the transformation to modify the original message by performing one or both of the following tasks:
```
